### DISCUSSION TITLE
Using Cloudinary for responsive images

### DISCUSSION THREAD

**Discussion by nhoizey** [11ty/eleventy#3965] [Enhancement Queue]
It would be great to be able to use [Cloudinary](https://cloudinary.com/) for responsive images.

For content images inside Markdown (or other formats), it could use presets and a Liquid syntax as in [the Jekyll-Cloudinary plugin](https://nhoizey.github.io/jekyll-cloudinary/).

It would be even better if it could use [native Markdown syntax](https://daringfireball.net/projects/markdown/syntax#img) and some "filters" in templates.

I already use Cloudinary, so I can help define the features, maybe even provide some code… ;-)

**Comment by eeeps** [comment]
I, for one, (am completely biased but) think this is a great idea! And would of course love to help make it a reality.

I had some simmilar thoughts when seeing this: https://twitter.com/zachleat/status/943502658295476225 ... 

@zachleat how exactly do you envision static asset handling things “layering on top?”

**Comment by zachleat** [comment]
I think this is going to be closely tied to what gets formulated for #13.

**Comment by Wilto** [comment]
![A monitor lizard, captioned with “I am monitoring this thread”](https://user-images.githubusercontent.com/439942/38430935-b8dfc6e2-3990-11e8-8920-86074a66581c.png)

**Comment by zachleat** [comment]
This repository is now using lodash style issue management for enhancements (see https://twitter.com/samselikoff/status/991395669016436736)

This means enhancement issues will now be closed instead of leaving them open. The enhancement backlog can be found here: https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+

**Comment by eeeps** [comment]
Now that we have shortcodes - here's a (very basic) start: https://github.com/eeeps/eleventy-respimg

**Comment by nhoizey** [comment]
I've made huge progress towards my aim to use standard Markdown syntax for images and yet generate responsive images. I'm using Eleventy transforms for that, enhancing the standard images HTML `<img src="…">` to `<img src="…" srcset="…" sizes="…">` and even embedding it in a `<figure>` if/when it makes sense.

I'll try to release it soon as a plugin that would support Cloudinary at first (I'm using it), but potentially other image CDNs, or even home made solutions.

Sorry for the teasing, but I wanted to let you know it will come soon, at last. 😁

**Comment by nhoizey** [comment]
Teasing is over, here's my plugin for simple responsive images in Eleventy:
https://github.com/nhoizey/eleventy-plugin-images-responsiver/

`eleventy-plugin-images-responsiver` is "just" a thin layer between Eleventy plugin and transformations system and [`images-responsiver`](https://github.com/nhoizey/images-responsiver), a generic HTML transformation Node.js module for simple responsive images.

Eleventy uses Markdown-it to transform Markdown into HTML, and then runs the transform added by `eleventy-plugin-images-responsiver`, which calls `images-responsiver` to actually transform the HTML.

Neither `eleventy-plugin-images-responsiver` nor `images-responsiver` transform (resize) the image files, there are multiple ways to do so: https://nhoizey.github.io/images-responsiver/tutorial-05-images-urls.html

**Comment by nhoizey** [comment]
On a [newer and simpler site](https://activites.nicolas-hoizey.com/), I'm now using the Eleventy Image plugin with Cloudinary.

It doesn't support advanced usage like my own plugin (different presets for example), but it's really simpler to use, I only have [these few lines](https://github.com/nhoizey/activites.nicolas-hoizey.com/blob/89eb70a38133372d5a997ad59e0524dd14600e34/eleventy.config.js#L35-L54) in my `eleventy.config.js` file:

```JavaScript
const imageOptions = {
  formats: ["jpeg"],
  widths: [640, 800],
  htmlOptions: {
    imgAttributes: {
      loading: "lazy",
      decoding: "async",
    },
    pictureAttributes: {}
  },
  fixOrientation: true,
}

if (process.env.CLOUDINARY_CLOUDNAME !== undefined) {
  imageOptions.urlFormat = ({ hash, src, width, format }) => {
    return `https://res.cloudinary.com/${process.env.CLOUDINARY_CLOUDNAME}/image/fetch/q_auto,f_auto/w_${width}/https://activites.nicolas-hoizey.com/${src.replace(/^src\//, "").replace(/^collections\//, "")}`;
  };
}

eleventyConfig.addPlugin(eleventyImageTransformPlugin, imageOptions);
```

`CLOUDINARY_CLOUDNAME` is an environment variable.

**Reply by eeeps** [reply]
I love this and thank you for sharing.

(Small note: it *might* be worth throwing a `c_limit` in with the `w_`, to prevent (server-side) upscaling)

**Reply by nhoizey** [reply]
@eeeps indeed, good catch! 👍
